platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetEventCreatorImpl.java

```diff
@@ -0,0 +1,392 @@
+
+        // Choose a random nerd, weighted by how much it is currently being bullied.
+
+        // First, figure out who is a nerd and sum up all bully scores.
+        int bullyScoreSum = 0;
+        final List<Integer> bullyScores = new ArrayList<>(possibleOtherParents.size());
+        for (final EventDescriptor nerd : possibleOtherParents) {
+            final int bullyScore = tipsetScoreCalculator.getBullyScoreForNode(nerd.getCreator());
+
+            final TipsetAdvancementWeight tipsetScore =
+                    tipsetScoreCalculator.getTheoreticalAdvancementScore(List.of(nerd));
+
+            if (bullyScore > 1 && tipsetScore.isNonZero()) {
+                // Note: if bully score is greater than 1, it is mathematically not possible
+                // for the advancement score to be zero. But in the interest in extreme caution,
+                // we check anyway, since it is very important never to create events with
+                // an advancement score of zero.
+
+                nerds.add(nerd);
+                bullyScores.add(bullyScore);
+                bullyScoreSum += bullyScore;
+            }
+        }
+
+        if (nerds.isEmpty()) {
+            // Note: this should be impossible, since we will not enter this method in the first
+            // place if there are no nerds. But better to be safe than sorry, and returning null
+            // is an acceptable way of saying "I can't create an event right now".
+            return null;
+        }
+
+        // Choose a random nerd.
+        final int choice = random.nextInt(bullyScoreSum);
+        int runningSum = 0;
+        for (int i = 0; i < nerds.size(); i++) {
+            runningSum += bullyScores.get(i);
+            if (choice < runningSum) {
+                final EventDescriptor nerd = nerds.get(i);
+                tipsetMetrics.getPityParentMetric(nerd.getCreator()).cycle();
+                return buildAndProcessEvent(nerd);
+            }
+        }
+
+        // This should be impossible.
+        throw new IllegalStateException("Failed to find an other parent");
+    }
+
+    /**
+     * Given an other parent, build the next self event and process it.
+     *
+     * @param otherParent the other parent, or null if there is no other parent
+     * @return the new event
+     */
+    private GossipEvent buildAndProcessEvent(@Nullable final EventDescriptor otherParent) {
+        final List<EventDescriptor> parentDescriptors = new ArrayList<>(2);
+        if (lastSelfEvent != null) {
+            parentDescriptors.add(lastSelfEvent);
+        }
+        if (otherParent != null) {
+            parentDescriptors.add(otherParent);
+        }
+
+        final GossipEvent event = assembleEventObject(otherParent);
+
+        final EventDescriptor descriptor = buildDescriptor(event);
+        tipsetTracker.addEvent(descriptor, parentDescriptors);
+        final TipsetAdvancementWeight score = tipsetScoreCalculator.addEventAndGetAdvancementScore(descriptor);
```

Comment:
more mixing of terms - "tipset score" vs "advancement score"

Reply:
Updated to make more consistent